app/utils/importHelpers.ts

```diff
@@ -489,27 +501,61 @@ async function importDuplicateWallets(
             addMessage
         );
 
-        const {
-            accountsOnIdentity,
-            credentialsOnIdentity,
-        } = findAccountsAndCredentialsOnIdentity(
+        await addAccountsAndCredentials(
             existingIdentity.id,
-            attachedAccounts,
-            attachedCredentials
+            identityInDatabase.id
         );
+    }
 
-        await insertNewAccountsAndCredentials(
-            identityInDatabase.id,
-            accountsOnIdentity,
-            credentialsOnIdentity,
-            addMessage
+    // We want to find the placeholder identities, whose real version are present.
+    const [
+        recoveredIdentities,
+        newIdentities,
+    ] = partition(nonExistingIdentities, (nonExistingIdentity) =>
+        isDuplicate(nonExistingIdentity, existingData.identities, [
+            'identityNumber',
+            'walletId',
+        ])
+    );
+    for (let i = 0; i < recoveredIdentities.length; i += 1) {
+        const importedIdentity = recoveredIdentities[i];
+
+        // Find the identity as it is in the database.
+        const existingIdentity = existingData.identities.find(
+            (ident) =>
+                ident.identityNumber === importedIdentity.identityNumber &&
+                ident.walletId === importedIdentity.walletId
         );
+
+        if (!existingIdentity) {
+            throw new Error(
+                'Internal error. An existing and matching identity should have been found, but was not.'
+            );
+        }
+
+        const existingId = existingIdentity.id;
+
+        if (importedIdentity.status !== IdentityStatus.Recovered) {
```

Comment:
It is confusing that we have the array `recoveredIdentities` that can contain identities where `status !== IdentityStatus.Recovered`. What other identities could be in `recoveredIdentities`?

Reply:
I have tried to refactor this and change some of names.